examples/erc1155/lib.rs

```diff
@@ -0,0 +1,792 @@
+            ensure!(token_id <= self.token_id_nonce, Error::UnexistentToken);
+
+            let caller = self.env().caller();
+            self.balances.insert((caller, token_id), value);
+
+            // Emit transfer event but with mint semantics
+            self.env().emit_event(TransferSingle {
+                operator: Some(caller),
+                from: None,
+                to: Some(caller),
+                token_id,
+                value,
+            });
+
+            Ok(())
+        }
+
+        // Helper function for performing single token transfers.
+        //
+        // Should not be used directly since it's missing certain checks which are important to the
+        // ERC-1155 standard (it is expected that the caller has already performed these).
+        fn perform_transfer(
+            &mut self,
+            from: AccountId,
+            to: AccountId,
+            token_id: TokenId,
+            value: Balance,
+            #[cfg_attr(test, allow(unused_variables))] data: Vec<u8>,
+        ) -> Result<()> {
+            let balance = self.balance_of(from, token_id);
+            ensure!(balance >= value, Error::InsufficientBalance);
+
+            self.balances
+                .entry((from, token_id))
+                .and_modify(|b| *b -= value);
+
+            self.balances
+                .entry((to, token_id))
+                .and_modify(|b| *b += value)
+                .or_insert(value);
+
+            let caller = self.env().caller();
+            self.env().emit_event(TransferSingle {
+                operator: Some(caller),
+                from: Some(from),
+                to: Some(from),
+                token_id,
+                value,
+            });
+
+            // This is disabled during tests due to the use of `eval_contract()` not being
+            // supported (tests end up panicking).
+            #[cfg(not(test))]
+            {
+                use ink_env::call::{
+                    build_call,
+                    utils::ReturnType,
+                    ExecutionInput,
+                    Selector,
+                };
+
+                // If our recipient is a smart contract we need to see if they accept or
+                // reject this transfer. If they reject it we need to revert the call.
+                let params = build_call::<Environment>()
+                    .callee(to)
+                    .gas_limit(5000)
+                    .exec_input(
+                        ExecutionInput::new(Selector::new(ON_ERC_1155_RECEIVED_SELECTOR))
+                            .push_arg(caller)
+                            .push_arg(from)
+                            .push_arg(token_id)
+                            .push_arg(value)
+                            .push_arg(data),
+                    )
+                    .returns::<ReturnType<Vec<u8>>>()
+                    .params();
+
+                match ink_env::eval_contract(&params) {
```

Comment:
I'm still trying to decide if these should be allowed to be gracefully handled by a contract or not. I'm leanings towards the "no" side right now

Reply:
`ReturnType` needed to be over `Result<Vec<u8>>` then, right?

Reply:
I've decided to not return a `Result` from `on_received()` - I don't want to risk the caller ignoring the `Err` case and getting user funds stuck